### PR TITLE
mgmt: ec_host_cmd: add logging system

### PR DESCRIPTION
**Pull request by niedzwiecki-dawid**
Add a native way to log Host Command communication.

Introduce a few logging levels. It allows adjusting the amount of log data depending on user needs - debug/normal usage.

Allow users to install a custom logging function.

### CODE REVIEW

File: subsys/mgmt/ec_host_cmd/Kconfig.logging [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	default 1
+	help
+	  Default Host Command logging level. The value should equals one of the enum
+	  in the enum ec_host_cmd_log_level.
+
+config EC_HOST_CMD_LOG_BUF_MAX
+	int "Max buffer to store request or response"
+	default 129
+	help
+	  Set maximum size of buffer to store request/response in hex. The amount of
```
Comment (fabiobaltieri) [inline subsys/mgmt/ec_host_cmd/Kconfig.logging:17]: "in bytes"? maybe it's obvious unsure...

File: subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c [hunk 2]
```diff
@@ -196,15 +271,15 @@ FUNC_NORETURN static void ec_host_cmd_thread(void *hc_handle, void *arg2, void *
 		/* Wait until RX messages is received on host interface */
 		k_sem_take(&rx->handler_owns, K_FOREVER);
 
+		ec_host_cmd_log_request(rx->buf);
 		status = verify_rx(rx);
 		if (status != EC_HOST_CMD_SUCCESS) {
-			send_status_response(hc->backend, tx, status);
+			ec_host_cmd_send_response(status, &args);
```
Comment (fabiobaltieri) [inline subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:277]: unrelated change?

File: subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c [hunk 3]
```diff
@@ -213,7 +288,7 @@ FUNC_NORETURN static void ec_host_cmd_thread(void *hc_handle, void *arg2, void *
 
 		/* No handler in this image for requested command */
 		if (found_handler == NULL) {
-			send_status_response(hc->backend, tx, EC_HOST_CMD_INVALID_COMMAND);
+			ec_host_cmd_send_response(EC_HOST_CMD_INVALID_COMMAND, &args);
```
Comment (fabiobaltieri) [inline subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:291]: same

File: subsys/mgmt/ec_host_cmd/Kconfig.logging [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+	  data sent via host command can be big and it would require a big buffer to
+	  print that in one piece. To print "n" number of data, a buffer with size =
+	  "2*n + 1" is required.
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/Kconfig.logging]: Done.

File: subsys/mgmt/ec_host_cmd/Kconfig.logging (hunk 1 above)
Comment (semihalf-niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/Kconfig.logging]: Done.

File: subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c [hunk 5]
```diff
@@ -161,10 +181,26 @@ int ec_host_cmd_send_response(enum ec_host_cmd_status status,
...
 	}
 
+	if (log_level >= EC_HOST_CMD_DEBUG_PARAMS && args->output_buf_size) {
+		size_t needed_hex_buf_len = 2 * args->output_buf_size + 1;
+		size_t hex_buf_len = MIN(needed_hex_buf_len, CONFIG_EC_HOST_CMD_LOG_BUF_MAX);
+		char hex_buf[hex_buf_len];
+
+		bin2hex(args->output_buf, (hex_buf_len - 1) / 2, hex_buf, hex_buf_len);
+		if (needed_hex_buf_len > CONFIG_EC_HOST_CMD_LOG_BUF_MAX) {
+			ec_host_cmd_print("HC resp:%s..", hex_buf);
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:199]: No secret meaning :-)

File: subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c [hunk 6]
```diff
@@ -174,6 +210,45 @@ int ec_host_cmd_send_response(enum ec_host_cmd_status status,
 	return hc->backend->api->send(hc->backend);
 }
 
+static void ec_host_cmd_log_request(const uint8_t *rx_buf)
+{
+	static uint16_t prev_cmd;
+	const struct ec_host_cmd_request_header *const rx_header = (void *)rx_buf;
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c]: Done.

File: subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c (hunk 2 above)
Comment (semihalf-niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:279]: It is related. I want to use one function to send all responses. It allows adding logging only to that function. I've created a separated commit for it.

File: subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c (hunk 3 above)
Comment (semihalf-niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:293]: Same as above.

**Review by fabiobaltieri — DISMISSED** [review]

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c [hunk 7]
```diff
@@ -161,10 +161,20 @@ int ec_host_cmd_send_response(enum ec_host_cmd_status status,
...
 	if (status != EC_HOST_CMD_SUCCESS) {
+		const struct ec_host_cmd_request_header *const rx_header =
+			(const struct ec_host_cmd_request_header *const)hc->rx_ctx.buf;
+
+		LOG_INF("HC 0x%04x err %d", rx_header->cmd_id, status);
 		send_status_response(hc->backend, tx, status);
 		return status;
 	}
 
+#ifdef CONFIG_EC_HOST_CMD_LOG_DBG_BUFFERS
```
Comment (keith-zephyr) [inline subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:172]: Non-blocking - can you use `if (IS_ENABLED(CONFG...))` here?

**Review by keith-zephyr — DISMISSED** [review]

File: subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c (hunk 7 above)
Comment (semihalf-niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:172]: Yes, done.

File: subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c [hunk 8]
```diff
@@ -174,6 +185,40 @@ int ec_host_cmd_send_response(enum ec_host_cmd_status status,
...
+	if (IS_ENABLED(CONFIG_EC_HOST_CMD_LOG_DBG_BUFFERS)) {
+		if (rx_header->data_len) {
+			const uint8_t *rx_data = rx_buf + RX_HEADER_SIZE;
+			static const char dbg_fmt[] = "HC 0x%04x.%d:";
+			/* Use sizeof because "%04x" needs 4 bytes for command id, and
+			 * %d needs 2 bytes for version, so no additional buffer is required.
+			 */
+			char dbg_raw[sizeof(dbg_fmt)];
+
+			sprintf(dbg_raw, dbg_fmt, rx_header->cmd_id, rx_header->cmd_ver);
```
Comment (fabiobaltieri) [inline subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:203]: Can you use snprintf? Just in case.

Comment (semihalf-niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c]: Sure, makes sense. Done.

**Review by fabiobaltieri — APPROVED** [review]

**Review by keith-zephyr — APPROVED** [review]
